src/stream/src/executor/over_window/frame_finder.rs

```diff
@@ -0,0 +1,302 @@
+        .unwrap();
+    let end = rows_frames
+        .iter()
+        .map(|bounds| bounds.end.n_following_rows())
+        .max_by(none_as_max_cmp)
+        .unwrap();
+
+    RowsFrameBounds {
+        start: start
+            .map(FrameBound::Preceding) // may produce Preceding(0), but doesn't matter
+            .unwrap_or(FrameBound::UnboundedPreceding),
+        end: end
+            .map(FrameBound::Following) // may produce Following(0), but doesn't matter
+            .unwrap_or(FrameBound::UnboundedFollowing),
+    }
+}
+
+/// For a `ROWS` frame, given a key in delta, find the cache key corresponding to
+/// the CURRENT ROW of the first frame that contains the given key.
+///
+/// ## Example
+///
+/// - Frame: `ROWS BETWEEN ? AND 2 FOLLOWING`
+/// - Cache: `[1, 2, 5]`
+/// - Delta: `[Delete 5]`
+///
+/// For delta key `5`, this function will return `1` as the *first curr key*.
+///
+/// More examples can be found in the comment inside [`find_curr_for_rows_frame`].
+pub(super) fn find_first_curr_for_rows_frame<'cache>(
+    frame_bounds: &'_ RowsFrameBounds,
+    part_with_delta: DeltaBTreeMap<'cache, CacheKey, OwnedRow>,
+    delta_key: &'cache CacheKey,
+) -> &'cache CacheKey {
+    find_curr_for_rows_frame::<true>(frame_bounds, part_with_delta, delta_key)
+}
+
+/// For a `ROWS` frame, given a key in delta, find the cache key corresponding to
+/// the CURRENT ROW of the last frame that contains the given key.
+///
+/// This is the symmetric function of [`find_first_curr_for_rows_frame`].
+pub(super) fn find_last_curr_for_rows_frame<'cache>(
+    frame_bounds: &'_ RowsFrameBounds,
+    part_with_delta: DeltaBTreeMap<'cache, CacheKey, OwnedRow>,
+    delta_key: &'cache CacheKey,
+) -> &'cache CacheKey {
+    find_curr_for_rows_frame::<false>(frame_bounds, part_with_delta, delta_key)
+}
+
+/// For a `ROWS` frame, given a key in `part_with_delta` corresponding to some
+/// CURRENT ROW, find the cache key corresponding to the start row in that frame.
+pub(super) fn find_frame_start_for_rows_frame<'cache>(
+    frame_bounds: &'_ RowsFrameBounds,
+    part_with_delta: DeltaBTreeMap<'cache, CacheKey, OwnedRow>,
+    curr_key: &'cache CacheKey,
+) -> &'cache CacheKey {
+    find_boundary_for_rows_frame::<true>(frame_bounds, part_with_delta, curr_key)
+}
+
+/// For a `ROWS` frame, given a key in `part_with_delta` corresponding to some
+/// CURRENT ROW, find the cache key corresponding to the end row in that frame.
+///
+/// This is the symmetric function of [`find_frame_start_for_rows_frame`].
+pub(super) fn find_frame_end_for_rows_frame<'cache>(
+    frame_bounds: &'_ RowsFrameBounds,
+    part_with_delta: DeltaBTreeMap<'cache, CacheKey, OwnedRow>,
+    curr_key: &'cache CacheKey,
+) -> &'cache CacheKey {
+    find_boundary_for_rows_frame::<false>(frame_bounds, part_with_delta, curr_key)
+}
+
+// -------------------------- ↑ PUBLIC INTERFACE ↑ --------------------------
+
+fn find_curr_for_rows_frame<'cache, const LEFT: bool>(
```

Comment:
Could you provide a docstring with the high-level semantics of this function? Will make it easier to review.

Reply:
```suggestion
/// Look at the function body for the semantics of the algorithm.
fn find_curr_for_rows_frame<'cache, const LEFT: bool>(
```